### PR TITLE
fix: gateway grpc client tls cert rotation

### PR DESCRIPTION
**Pull request by RepoRonin**
Fix HTTP gateway gRPC client not refreshing TLS credentials after certificate rotation

#### What problem is being solved?

When TLS certificate rotation occurs (e.g. via cert-manager), the HTTP gateway becomes unable to proxy requests to the internal gRPC server, causing all HTTP API calls to fail with a 500 and the error:

```
connection error: desc = "transport: authentication handshake failed: tls: failed to verify certificate: x509: certificate signed by unknown authority"
```

The gRPC server correctly uses [watchAndLoadCertificateWithCertWatcher](cci:1://file:///home/shashank/PROJECTS/openfga/cmd/run/run.go:1060:0-1088:1) (via controller-runtime's certwatcher) to dynamically reload the serving certificate. However, the HTTP gateway's internal gRPC client loads its TLS credentials once at startup using `credentials.NewClientTLSFromFile` and never refreshes them. When the server starts presenting the new certificate, the client still holds the old cert as its trust anchor, causing the TLS handshake to fail.

#### How is it being solved?

The static `credentials.NewClientTLSFromFile` call in [dialGrpc](cci:1://file:///home/shashank/PROJECTS/openfga/cmd/run/run.go:609:0-670:1) is replaced with the existing [watchAndLoadCertificateWithCertWatcher](cci:1://file:///home/shashank/PROJECTS/openfga/cmd/run/run.go:1060:0-1088:1) helper, which uses controller-runtime's certwatcher to watch the certificate file for changes. A `VerifyPeerCertificate` callback is used to dynamically build a cert pool from the watcher's latest certificate on each new connection and verify the server's presented certificate against it. This ensures the gateway client trusts the new server certificate immediately after rotation without requiring a restart.

#### What changes are made to solve it?

- [dialGrpc](cci:1://file:///home/shashank/PROJECTS/openfga/cmd/run/run.go:609:0-670:1) is updated to accept a `context.Context` (required to run the certwatcher goroutine) and uses `VerifyPeerCertificate` with the dynamically watched certificate as a trust anchor instead of a statically loaded credential.
- The call site in [Run](cci:1://file:///home/shashank/PROJECTS/openfga/cmd/run/run.go:780:0-1016:1) is updated to pass the server context to [dialGrpc](cci:1://file:///home/shashank/PROJECTS/openfga/cmd/run/run.go:609:0-670:1).
- A new test [TestHTTPServerWithGRPCTLSEnabled](cci:1://file:///home/shashank/PROJECTS/openfga/cmd/run/run_test.go:791:0-834:1) is added, which starts the server with gRPC TLS enabled and the HTTP gateway active, then verifies that HTTP requests can be proxied through the gateway successfully.

## References

- closes https://github.com/openfga/openfga/issues/2948

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

## Release Notes

* **New Features**
  * Enhanced gRPC TLS security with dynamic certificate rotation, enabling certificates to be updated at runtime without requiring a service restart.

* **Tests**
  * Added test coverage for gRPC TLS-enabled scenarios to ensure secure server connections function properly.

### CODE REVIEW

**Comment by senojj** [comment]
A few thoughts on this:
1. We should avoid creating a new pool on every connection. Instead, a global pool that is managed behind an `atomic.Pointer`, and updated by a watcher whenever the cert file changes, is the most ideal.
2. Only the leaf certificate verification really matters in this case, as it is a local cert with a local connection, and an exact match is the target.
3. The hostname _must_ be verified when taking this approach, as Go disables automatic hostname checks when InsecureSkipVerify is set to `true` in the TLS configuration.

**Comment by RepoRonin** [comment]
I am gonna look into it, and fix this

**Comment by RepoRonin** [comment]
@senojj I made the changes, do inform me on further improvements

**Comment by senojj** [comment]
You should add the leaf certificate to the RootCA pool and use the pool to verify the server's certificate; essentially certificate pinning. Just comparing the bytes of the certificate is insecure.

Also you can verify using the DNS name and the pool at the same time.

This PR adds a lot of complexity to simply add behavior to update the trusted certificate. The watcher has the ability to register a callback that is triggered on file changes.

**Comment by RepoRonin** [comment]
## Changes

Previously, `VerifyPeerCertificate` in [dialGrpc](cci:1://file:///home/shashank/PROJECTS/openfga/cmd/run/run.go:610:0-668:1) was rebuilding an `x509.CertPool`
on every TLS handshake, allocating and re-parsing certificates per connection.

I replaced this with a single pool held behind an `atomic.Pointer[x509.CertPool]`,
managed by a [newAtomicCertPool](cci:1://file:///home/shashank/PROJECTS/openfga/cmd/run/run.go:670:0-711:1) helper that:

- Creates a `certwatcher` instance and uses [RegisterCallback](cci:1://file:///home/shashank/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.23.1/pkg/certwatcher/certwatcher.go:91:0-100:1) to build and atomically
  swap in a new pinned pool (containing only the leaf certificate) whenever the cert
  file changes on disk — no polling goroutine needed.
- Exposes the pool as an atomic pointer, so each TLS handshake does a single [Load()](cci:1://file:///home/shashank/PROJECTS/openfga/cmd/run/run.go:1101:0-1129:1)
  with no allocation.

`VerifyPeerCertificate` now does a single `peerCert.Verify` call with both `Roots`
(the pinned pool) and `DNSName` (hostname extracted from the gRPC address), handling
certificate pinning and hostname verification together — the latter being required
since `InsecureSkipVerify: true` disables Go's automatic hostname check.

@senojj do tell me if further changes are required

**Comment by senojj** [comment]
A second watcher, watching the gRPC certificate file is unnecessary. Utilize the same watcher that is currently in use.

**Comment by RepoRonin** [comment]
> A second watcher, watching the gRPC certificate file is unnecessary. Utilize the same watcher that is currently in use.

@senojj did the change, anything else ?

**Comment by senojj** [comment]
Every call to `dialGrpc` will overwrite the grpc watcher's callback. Currently the function is only called once, however this could change in the future. 

Creating a package-level global pool would allow an "update" callback function to be registered within the `watchAndLoadCertificateWithCertWatcher` function without needing to return the watcher and pass it around throughout the codebase. Additionally, each connection created via `dialGrpc` would use the same global pool.

**Comment by RepoRonin** [comment]
> Every call to `dialGrpc` will overwrite the grpc watcher's callback. Currently the function is only called once, however this could change in the future.
> 
> Creating a package-level global pool would allow an "update" callback function to be registered within the `watchAndLoadCertificateWithCertWatcher` function without needing to return the watcher and pass it around throughout the codebase. Additionally, each connection created via `dialGrpc` would use the same global pool.

Could you help me with that, like how do I solve it ?

**Comment by RepoRonin** [comment]
> Every call to `dialGrpc` will overwrite the grpc watcher's callback. Currently the function is only called once, however this could change in the future.
> 
> Creating a package-level global pool would allow an "update" callback function to be registered within the `watchAndLoadCertificateWithCertWatcher` function without needing to return the watcher and pass it around throughout the codebase. Additionally, each connection created via `dialGrpc` would use the same global pool.

check now

File: cmd/run/run.go [hunk 1]
```diff
@@ -1024,10 +1063,20 @@ func watchAndLoadCertificateWithCertWatcher(ctx context.Context, certPath, keyPa
...
+		leaf, parseErr := x509.ParseCertificate(cert.Certificate[0])
+		if parseErr != nil {
+			logger.Error("failed to parse rotated certificate for cert pool", zap.Error(parseErr))
+			return
+		}
+		pool := x509.NewCertPool()
+		pool.AddCert(leaf)
+		grpcTLSCertPool.Store(pool)
+		logger.Info("gRPC client cert pool updated", zap.String("certPath", certPath))
+	})
```
Comment (senojj) [inline cmd/run/run.go]: `watchAndLoadCertificateWithCertWatcher` is also called by the HTTP server code. We do not want this callback to be applied for the HTTP server's certificate. The callback can be a package-level function, instead of a closure (to avoid unnecessary allocation), and should be passed into the `watchAndLoadCertificateWithCertWatcher` function as an argument, only for the grpc server setup. For the HTTP server setup, a `nil` value should be passed as the callback value.

Comment (RepoRonin) [reply cmd/run/run.go]: I did the changes, check now

File: cmd/run/run.go [hunk 2]
```diff
@@ -82,6 +84,27 @@ const (
 	datastoreURIFlag    = "datastore-uri"
 )
 
+// grpcTLSCertPool is a package-level pinned certificate pool for verifying
+// the gRPC server's TLS certificate from the HTTP gateway client. It is
+// populated and kept up-to-date by watchAndLoadCertificateWithCertWatcher
+// via the certwatcher's RegisterCallback. Every connection created by
+// dialGrpc shares this single pool.
+var grpcTLSCertPool sync.Pointer[x509.CertPool]
```
Comment (senojj) [inline cmd/run/run.go:92]: For safety, initialize the `grpcTLSCertPool` to an empty `x509.CertPool` on package `init`.

File: cmd/run/run.go [hunk 3]
```diff
@@ -82,6 +84,27 @@ const (
...
+
+// updateGRPCCertPool builds a pinned x509.CertPool from the leaf certificate
+// and stores it in the global grpcTLSCertPool. It is intended to be passed as
+// the callback argument to watchAndLoadCertificateWithCertWatcher for the gRPC
+// server setup.
+func updateGRPCCertPool(cert tls.Certificate) {
+	leaf, err := x509.ParseCertificate(cert.Certificate[0])
```
Comment (senojj) [inline cmd/run/run.go:109]: For safety, ensure that `len(cert.Certificate) > 0` _before_ indexing.

File: cmd/run/run_test.go [hunk 4]
```diff
@@ -789,6 +789,51 @@ func TestGRPCServingTLS(t *testing.T) {
...
+		defer certsAndKeys.Clean()
+
+		cfg := testutils.MustDefaultConfigWithRandomPorts()
+		cfg.GRPC.TLS = &serverconfig.TLSConfig{
+			Enabled:  true,
+			CertPath: certsAndKeys.serverCertFile,
+			KeyPath:  certsAndKeys.serverKeyFile,
+		}
+		// Port for TLS cannot be 0.0.0.0
+		cfg.GRPC.Addr = strings.ReplaceAll(cfg.GRPC.Addr, "0.0.0.0", "localhost")
```
Comment (senojj) [inline cmd/run/run_test.go]: Must be "localhost" because created test cert is for that domain name.

File: cmd/run/run_test.go [hunk 5]
```diff
@@ -789,6 +789,51 @@ func TestGRPCServingTLS(t *testing.T) {
 	})
 }
 
+func TestHTTPServerWithGRPCTLSEnabled(t *testing.T) {
```
Comment (senojj) [inline cmd/run/run_test.go:801]: This test does not exercise the certificate rotation path, so need another to do so.

File: cmd/run/run_test.go [hunk 6]
```diff
@@ -789,6 +789,51 @@ func TestGRPCServingTLS(t *testing.T) {
...
+		// credentials (via certwatcher) rather than a one-time static load.
+		client := retryablehttp.NewClient()
+		t.Cleanup(client.HTTPClient.CloseIdleConnections)
+
+		resp, err := client.Get(fmt.Sprintf("http://%s/stores", cfg.HTTP.Addr))
+		require.NoError(t, err)
+		defer resp.Body.Close()
+		require.Equal(t, http.StatusOK, resp.StatusCode)
+	})
+}
```
[the remark this answers is not in the document]
Comment (senojj) [reply cmd/run/run_test.go:901]: The implementation now uses the leaf certificate of the same certificate used by the server as the root CA. When the server's certificate is verified, the leaf cert is used. The intent is to pin the certificate.

**Comment by RepoRonin** [comment]
> Starting to shape up much better. Some additional changes needed though.

yes, working on them.
Thanks for the feedback

**Comment by RepoRonin** [comment]
> Starting to shape up much better. Some additional changes needed though.

Hi ! I did the changes. Let me know on the follow ups

File: cmd/run/run_test.go (hunk 4 above)
Comment (senojj) [reply cmd/run/run_test.go]: Your AI prompt resulted in multiple changed comments, but does not enforce that the value here is _always_ "localhost".

**Comment by senojj** [comment]
Additionally, you will need to sign the CLA before this PR can be considered for merging, and you must add a "Changed" entry to the changelog [see here](https://github.com/openfga/openfga/blob/main/CHANGELOG.md).

File: cmd/run/run_test.go (hunk 4 above)
Comment (RepoRonin) [reply cmd/run/run_test.go]: Yes, i used AI cuz i could not understand this one more clearly, isn't it already "localhost" .
Could you guide me more on what needs to be done ?

**Comment by RepoRonin** [comment]
> Additionally, you will need to sign the CLA before this PR can be considered for merging, and you must add a "Changed" entry to the changelog [see here](https://github.com/openfga/openfga/blob/main/CHANGELOG.md).

I have signed the CLA, & added the "Changed" entry

File: cmd/run/run_test.go (hunk 4 above)
Comment (senojj) [reply cmd/run/run_test.go]: This code changes the value to "localhost" only when the current value is "0.0.0.0". If the default value of `cfg.GRPC.Addr` were something like "test.com", this test would fail.

It is fine to use AI to generate the code changes, but it is important that you be able to properly vet, and understand all of the code that it generates before submitting it. I only mention this because much of the back-and-forth in this review was due to AI generated code changes that lacked a complete understanding of the change.

File: cmd/run/run_test.go [hunk 7]
```diff
@@ -703,8 +712,6 @@ func TestHTTPServingTLS(t *testing.T) {
 			CertPath: certsAndKeys.serverCertFile,
 			KeyPath:  certsAndKeys.serverKeyFile,
 		}
-		// Port for TLS cannot be 0.0.0.0
-		cfg.HTTP.Addr = strings.ReplaceAll(cfg.HTTP.Addr, "0.0.0.0", "localhost")
```
Comment (senojj) [inline cmd/run/run_test.go:707]: This code is unrelated to your code change. Removing it will now cause the test to fail when `cfg.HTTP.Addr` defaults to anything other than "localhost". It currently defaults to "0.0.0.0", so it will currently fail.

File: cmd/run/run_test.go [hunk 8]
```diff
@@ -769,8 +776,6 @@ func TestGRPCServingTLS(t *testing.T) {
 			CertPath: certsAndKeys.serverCertFile,
 			KeyPath:  certsAndKeys.serverKeyFile,
 		}
-		// Port for TLS cannot be 0.0.0.0
-		cfg.GRPC.Addr = strings.ReplaceAll(cfg.GRPC.Addr, "0.0.0.0", "localhost")
```
Comment (senojj) [inline cmd/run/run_test.go:773]: This code is unrelated to your code change. Removing it will now cause the test to fail when `cfg.GRPC.Addr` defaults to anything other than "localhost". It currently defaults to "0.0.0.0", so it will currently fail.

File: cmd/run/run_test.go [hunk 9]
```diff
@@ -789,6 +794,106 @@ func TestGRPCServingTLS(t *testing.T) {
...
+			if err := runServer(ctx, cfg); err != nil {
+				log.Fatal(err)
+			}
+		}()
+
+		certPool := x509.NewCertPool()
+		certPool.AddCert(certsAndKeys.caCert)
+		creds := credentials.NewClientTLSFromCert(certPool, "")
+
+		testutils.EnsureServiceHealthy(t, cfg.GRPC.Addr, cfg.HTTP.Addr, creds)
```
Comment (senojj) [inline cmd/run/run_test.go:830]: this will fail if `cfg.GRPC.Addr` is anything other than "localhost". It currently defaults to "0.0.0.0". My previous comment provided instruction to use the "localhost" value directly, or set `cfg.GRPC.Addr` explicitly.

So simply adding a line like `cfg.GRPC.Addr = "localhost"` would have been sufficient.

File: cmd/run/run_test.go [hunk 10]
```diff
@@ -789,6 +794,106 @@ func TestGRPCServingTLS(t *testing.T) {
...
```
Comment (senojj) [inline cmd/run/run_test.go:869]: this will fail if cfg.GRPC.Addr is anything other than "localhost". It currently defaults to "0.0.0.0".

File: CHANGELOG.md [hunk 11]
```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 Try to keep listed changes to a concise bulleted list of simple explanations of changes. Aim for the amount of information needed so that readers can understand where they would look in the codebase to investigate the changes' implementation, or where they would look in the documentation to understand how to make use of the change in practice - better yet, link directly to the docs and provide detailed information there. Only elaborate if doing so is required to avoid breaking changes or experimental features from ruining someone's day.
 
 ## [Unreleased]
+### Changed
+- HTTP gateway's internal gRPC client now uses dynamic TLS credentials that automatically update on certificate rotation via certwatcher, preventing connection failures when certificates are rotated (e.g., by cert-manager). [#XXXX](https://github.com/openfga/openfga/pull/XXXX)
```
Comment (senojj) [inline CHANGELOG.md]: replace "XXXX" with this PR's number.

Comment (RepoRonin) [reply CHANGELOG.md]: done

**Comment by RepoRonin** [comment]
@senojj check now

**Comment by senojj** [comment]
I'm honestly unsure what the latest changes are attempting to achieve. The AI tool that you are using is removing and then adding back and removing and then adding back. These recent changes still fail to always enforce "localhost" as a value for the gRPC server, explicitly.

If the ask is too unclear, please feel free to ask further clarifying questions.

Alternatively, if you are familiar with programming, I would suggest writing this minimal change by hand.

If you are not familiar with programming, then I would suggest that we close this PR, and I can implement the enhancement within another branch.

File: CHANGELOG.md (hunk 11 above)
Comment (senojj) [reply CHANGELOG.md]: You did not make the change.

**Comment by RepoRonin** [comment]
> I'm honestly unsure what the latest changes are attempting to achieve. The AI tool that you are using is removing and then adding back and removing and then adding back. These recent changes still fail to always enforce "localhost" as a value for the gRPC server, explicitly.
> 
> If the ask is too unclear, please feel free to ask further clarifying questions.
> 
> Alternatively, if you are familiar with programming, I would suggest writing this minimal change by hand.
> 
> If you are not familiar with programming, then I would suggest that we close this PR, and I can implement the enhancement within another branch.

I'll work on it.

**Comment by RepoRonin** [comment]
@senojj 
I'd like your review on this : 
where we bind the gRPC listener if I replace this : 


listener, err := net.Listen("tcp", cfg.GRPC.Addr)


with 

```go
host, port, err := net.SplitHostPort(cfg.GRPC.Addr)
if err != nil {
    return fmt.Errorf("invalid grpc addr: %w", err)
}

_ = host // ignore whatever user passed

cfg.GRPC.Addr = fmt.Sprintf("localhost:%s", port)

listener, err := net.Listen("tcp", cfg.GRPC.Addr)

**Comment by senojj** [comment]
> @senojj I'd like your review on this : where we bind the gRPC listener if I replace this :
> 
> listener, err := net.Listen("tcp", cfg.GRPC.Addr)
> 
> with
> 
> ```go
> host, port, err := net.SplitHostPort(cfg.GRPC.Addr)
> if err != nil {
>     return fmt.Errorf("invalid grpc addr: %w", err)
> }
> 
> _ = host // ignore whatever user passed
> 
> cfg.GRPC.Addr = fmt.Sprintf("localhost:%s", port)
> 
> listener, err := net.Listen("tcp", cfg.GRPC.Addr)
> ```

Remember this is just for a test. For our needs we need only specify a valid hostname, given our TLS certificate, and a usable port.

Therefore something like this would suffice:
```go
cfg.GRPC.Addr = "localhost:8082"
```
If you wanted to get fancier with a port scanning function, feel free.

**Comment by RepoRonin** [comment]
> > @senojj I'd like your review on this : where we bind the gRPC listener if I replace this :
> > listener, err := net.Listen("tcp", cfg.GRPC.Addr)
> > with
> > ```go
> > host, port, err := net.SplitHostPort(cfg.GRPC.Addr)
> > if err != nil {
> >     return fmt.Errorf("invalid grpc addr: %w", err)
> > }
> > 
> > _ = host // ignore whatever user passed
> > 
> > cfg.GRPC.Addr = fmt.Sprintf("localhost:%s", port)
> > 
> > listener, err := net.Listen("tcp", cfg.GRPC.Addr)
> > ```
> 
> Remember this is just for a test. For our needs we need only specify a valid hostname, given our TLS certificate, and a usable port.
> 
> Therefore something like this would suffice:
> 
> ```go
> cfg.GRPC.Addr = "localhost:8082"
> ```
> 
> If you wanted to get fancier with a port scanning function, feel free.

Makes sense, since this is only for testing and we just need a valid hostname that matches the TLS certificate and a free port, I’ll simplify this to explicitly set cfg.GRPC.Addr = "localhost:8082" and avoid the additional parsing logic.

**Comment by RepoRonin** [comment]
@senojj 
give it a check now

File: CHANGELOG.md (hunk 11 above)
Comment (senojj) [reply CHANGELOG.md]: This still need to be corrected.

Comment (RepoRonin) [reply CHANGELOG.md]: > This still need to be corrected.

done

File: cmd/run/run.go [hunk 12]
```diff
@@ -612,10 +645,40 @@ func (s *ServerContext) dialGrpc(config *serverconfig.Config) *grpc.ClientConn {
...
+				if err != nil {
+					return fmt.Errorf("failed to parse peer certificate: %w", err)
+				}
+
+				// Verify the leaf certificate against the global pinned pool
+				// and check the hostname in a single call. The pool is
+				// updated atomically by the watcher callback on cert rotation.
+				_, err = peerCert.Verify(x509.VerifyOptions{
+					Roots:   grpcTLSCertPool.Load(),
+					DNSName: grpcHost,
```
[the remark this answers is not in the document]
Comment (senojj) [reply cmd/run/run.go:674]: The server's leaf certificate is added to the client's RootCAs pool in order to pin the certificate.

File: cmd/run/run.go [hunk 13]
```diff
@@ -82,6 +84,37 @@ const (
...
+func updateGRPCCertPool(cert tls.Certificate) {
+	if len(cert.Certificate) == 0 {
+		return
+	}
+	leaf, err := x509.ParseCertificate(cert.Certificate[0])
+	if err != nil {
+		return
+	}
+	pool := x509.NewCertPool()
+	pool.AddCert(leaf)
```
[the remark this answers is not in the document]
Comment (senojj) [reply cmd/run/run.go:114]: This PR is using certificate pinning as suggested.

**Review by senojj — APPROVED** [review]
The code and implementation look good to me. The rest of the team will review on Monday.

**Comment by RepoRonin** [comment]
HI @senojj 
Currently this codecov/patch is failing, I don't know about it, could you tell me about it ? 
How do I resolve this ?

**Comment by senojj** [comment]
> Currently this codecov/patch is failing, I don't know about it, could you tell me about it ?
How do I resolve this?

This is a check for code test coverage. What it is saying is that, of the lines of code you changed, 66.67% of those lines were hit during the test runs. The lines that are not covered by unit tests are primarily error edge cases that are very unlikely to happen.

This is not a required check, however, if you wish to satisfy it, I would not oppose you doing so, so long as you keep the tests simple and clean.

**Comment by RepoRonin** [comment]
> The code and implementation look good to me. The rest of the team will review on Monday.

Any updates ?

**Comment by RepoRonin** [comment]
Glad I could help
